k2/csrc/context.cu

```diff
@@ -4,118 +4,42 @@
-  DeviceType GetDeviceType() const override { return kCpu; }
-
-  void *Allocate(std::size_t bytes, void **deleter_context) override {
-    void *p = nullptr;
-    if (bytes) {
-      int32_t ret = posix_memalign(&p, kAlignment, bytes);
-      K2_DCHECK_EQ(ret, 0);
-    }
-    if (deleter_context) *deleter_context = nullptr;
-    return p;
-  }
-
-  bool IsCompatible(const Context &other) const override {
-    return other.GetDeviceType() == kCpu;
-  }
-
-  void Deallocate(void *data, void * /*deleter_context*/) override {
-    free(data);
-  }
-};
-
-class CudaContext : public Context {
- public:
-  explicit CudaContext(int32_t gpu_id) : gpu_id_(gpu_id) {
-    if (gpu_id_ != -1) {
-      auto ret = cudaSetDevice(gpu_id_);
-      K2_CHECK_CUDA_ERROR(ret);
-    }
-    // TODO(haowen): choose one from available GPUs if gpu_id == -1?
-    // and handle GPU ids from multiple machines.
-    auto ret = cudaStreamCreate(&stream_);
-    K2_CHECK_CUDA_ERROR(ret);
-  }
-  ContextPtr GetCpuContext() override { return nullptr; }
-  ContextPtr GetPinnedContext() override { return nullptr; }
-  DeviceType GetDeviceType() const override { return kCuda; }
-  int GetDeviceId() const override { return gpu_id_; }
-
-  void *Allocate(std::size_t bytes, void **deleter_context) override {
-    void *p = nullptr;
-    if (bytes) {
-      auto ret = cudaMalloc(&p, bytes);
-      K2_CHECK_CUDA_ERROR(ret);
-    }
-    if (deleter_context) *deleter_context = nullptr;
-    return p;
-  }
-
-  bool IsCompatible(const Context &other) const override {
-    return other.GetDeviceType() == kCuda && other.GetDeviceId() == gpu_id_;
-  }
-
-  void Deallocate(void *data, void * /*deleter_context*/) override {
-    auto ret = cudaFree(data);
-    K2_CHECK_CUDA_ERROR(ret);
-  }
-
-  cudaStream_t GetCudaStream() const override { return stream_; }
-
-  void Sync() const override {
-    auto ret = cudaStreamSynchronize(stream_);
-    K2_CHECK_CUDA_ERROR(ret);
-  }
-
-  ~CudaContext() {
-    auto ret = cudaStreamDestroy(stream_);
-    K2_CHECK_CUDA_ERROR(ret);
-  }
-
- private:
-  int32_t gpu_id_;
-  cudaStream_t stream_;
-};
-
-ContextPtr GetCpuContext() { return std::make_shared<CpuContext>(); }
-
 ContextPtr GetCudaContext(int32_t gpu_id /*= -1*/) {
   return std::make_shared<CudaContext>(gpu_id);
 }
 
-RegionPtr NewRegion(ContextPtr &context, std::size_t num_bytes) {
-  // .. fairly straightforward.  Sets bytes_used to num_bytes, caller can
-  // overwrite if needed.
-  auto ans = std::make_shared<Region>();
-  ans->context = context->shared_from_this();
-  // TODO(haowen): deleter_context is always null with above constructor,
-  // we need add another constructor of Region to allow the caller
-  // to provide deleter_context.
-  ans->data = context->Allocate(num_bytes, &ans->deleter_context);
-  ans->num_bytes = num_bytes;
-  ans->bytes_used = num_bytes;
-  return ans;
+template <typename LambdaT>
+__global__ void eval_lambda(int32_t n, LambdaT lambda) {
```

Comment:
No, this needs to be in a .h or .cuh file because it's templated.  
I'd rather name all the header files .h, actually, and all the .cc files .cu, because actually code that doesn't explicitly use a GPU, often does implicitly due to templates.

Reply:
To distinguish .cu and .cc help us avoid call other units' code, as that would need the separate compilation mode and make cmake hard to work fine and tune. While separate compilation mode degrades performance at runtime.

Reply:
About template, it need discuss about cuda could or not use template in .cu. As nvcc treats .h differ with c++, in this situation that it gets compiled and gtest passed, we need make it sure. Otherwise, to put __global__ implementation code into *_inl.cuh, actually is not that right. Need to make sure.